tools/rpkg/R/backend-dbplyr__duckdb_connection.R

```diff
@@ -0,0 +1,383 @@
+}
+
+#' Declare which version of dbplyr API is being called.
+#' @param con A \code{\link{dbConnect}} object, as returned by \code{dbConnect()}
+#' @export
+dbplyr_edition.duckdb_connection <- function(con) {
+  2L
+}
+
+#' Description of the database connection
+#' @param con A \code{\link{dbConnect}} object, as returned by \code{dbConnect()}
+#' @name db_connection_describe
+#' @return
+#' String consisting of DuckDB version, user login name, operating system, R version and the name of database
+#' @export
+db_connection_describe.duckdb_connection <- function(con) {
+  info <- DBI::dbGetInfo(con)
+  paste0(
+    "DuckDB ", info$db.version, " [", Sys.info()["login"], "@",
+    paste(Sys.info()[c("sysname", "release")], collapse = " "), ":",
+    "R ", R.version$major, ".", R.version$minor, "/", info$dbname, "]"
+  )
+}
+
+duckdb_grepl <- function(pattern, x, ignore.case = FALSE, perl = FALSE, fixed = FALSE, useBytes = FALSE) {
+  # https://duckdb.org/docs/sql/functions/patternmatching
+  if (any(c(perl, fixed, useBytes))) {
+    stop("Parameters `perl`, `fixed` and `useBytes` in grepl are not currently supported in DuckDB backend", .call = FALSE)
+  }
+
+  sql_expr <- pkg_method("sql_expr", "dbplyr")
+
+  if (ignore.case) {
+    icpattern <- paste0("(?i)", pattern)
+    sql_expr(REGEXP_MATCHES((!!x), (!!icpattern)))
+  } else {
+    sql_expr(REGEXP_MATCHES((!!x), (!!pattern)))
+  }
+}
+
+
+#' Customized translation functions for DuckDB SQL
+#' @param con A \code{\link{dbConnect}} object, as returned by \code{dbConnect()}
+#' @name sql_translation
+#' @export
+sql_translation.duckdb_connection <- function(con) {
+  sql_variant <- pkg_method("sql_variant", "dbplyr")
+  sql_translator <- pkg_method("sql_translator", "dbplyr")
+  build_sql <- pkg_method("build_sql", "dbplyr")
+  sql_expr <- pkg_method("sql_expr", "dbplyr")
+  sql_prefix <- pkg_method("sql_prefix", "dbplyr")
+  sql_cast <- pkg_method("sql_cast", "dbplyr")
+  sql_paste <- pkg_method("sql_paste", "dbplyr")
+  sql_aggregate <- pkg_method("sql_aggregate", "dbplyr")
+  sql_aggregate_2 <- pkg_method("sql_aggregate_2", "dbplyr")
+  win_aggregate <- pkg_method("win_aggregate", "dbplyr")
+  win_aggregate_2 <- pkg_method("win_aggregate_2", "dbplyr")
+  win_over <- pkg_method("win_over", "dbplyr")
+  win_current_order <- pkg_method("win_current_order", "dbplyr")
+  win_current_group <- pkg_method("win_current_order", "dbplyr")
+
+
+  base_scalar <- pkg_method("base_scalar", "dbplyr")
+  base_agg <- pkg_method("base_agg", "dbplyr")
+  base_win <- pkg_method("base_win", "dbplyr")
+
+  sql_variant(
+    sql_translator(
+      .parent = base_scalar,
+      as.raw = sql_cast("VARBINARY"),
+      `%%` = function(a, b) build_sql("(CAST((", a, ") AS INTEGER)) % (CAST((", b, ") AS INTEGER))"),
+      `%/%` = function(a, b) build_sql("(((CAST((", a, ") AS INTEGER)) - (CAST((", a, ") AS INTEGER)) % (CAST((", b, ") AS INTEGER)))) / (CAST((", b, ") AS INTEGER))"),
```

Comment:
Why is this so complicated?

Reply:
For the modulo R accepts non-integer values, but those give errors in the database, so extra cast is required here. I may simplify by using ::-notation for casts here.

For the integer division there seemed no to be a ready made function available in DuckDB, so I just solved %/% from the guaranteed equation `x == (x %% y) + y * ( x %/% y )`. If there is any more direct way to calculate that, it certainly can be changed.

Reply:
It seems that integer division remains problematic (tidyverse/dbplyr#108), so I commented out the current version. This probably requires feature request for DuckDB to get an additional function that would give equivalent results with R.

Reply:
I opened an issue for alternative modulus and integer division functions (#2981).

Reply:
It seems that the required correction for modulus can be done as follows. Would that be OK (until the alternative functions will be implemented in DuckDB)?

``` r
df <- dplyr::tibble(
  x = c(10, 10, -10, -10), 
  y = c(3, -3, 3, -3)
)
con <- DBI::dbConnect(duckdb::duckdb(), ":memory:")
dplyr::copy_to(con,df)
DBI::dbGetQuery(con,'SELECT "x", "y", "x" % "y" AS "mod", CASE WHEN ("y" < 0.0 AND "x" % "y" > 0.0) OR ("y" > 0.0 AND "x" % "y" < 0.0) THEN ("x" % "y" + "y") ELSE ("x" % "y") END AS "fmod" FROM "df"')
#>     x  y mod fmod
#> 1  10  3   1    1
#> 2  10 -3   1   -2
#> 3 -10  3  -1    2
#> 4 -10 -3  -1   -1
```

<sup>Created on 2022-01-25 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

Reply:
Well, it appears that in R `integer division(x,y)` is simply `floor(x/y)` and `modulus(x,y)` is `x-y*floor(x/y)`. I will change the translations correspondingly.

Reply:
Simplified versions + tests added.